File: ImageProsessing/m_ImageRestoration/ILL_Posed/Deblur_DL/Black-box image deblurring and defiltering/mRL.m
function [S, E] = mRL(F, y, options)
% Derivative free variant of Lucy-Richardson

options.null = 0;
maxiter = getoptions(options, 'maxiter', 500);

S = y;

E = [];
e = img_norm(y - F(S));
E = [E; e];

for i=1:maxiter
    r1 = F(S)./(abs(y)+eps);
    r1 = r1(end:-1:1,end:-1:1,:);
    r2 = F(r1);
    r2 = r2(end:-1:1,end:-1:1,:);
    S = S./(abs(r2)+eps);
    
    r1 = y./(abs(F(S))+eps);
    r1 = r1(end:-1:1,end:-1:1,:);
    r2 = F(r1);
    r2 = r2(end:-1:1,end:-1:1,:);
    S = S.*r2;
    
    e = img_norm(y - F(S));
    E = [E; e];
end

end

File: ImageProsessing/p_ImageRestoration/ILL_Posed/Deblur_DL/Black-box image deblurring and defiltering/src/mRL.py
# -*- coding: utf-8 -*-

import numpy as np
import utils


"""## Modified Richardson-Lucy"""

def mRL(F, y, maxiter=500):
    RL = y.copy()

    for i in range(maxiter):
        r1 = F(RL)/(np.abs(y)+1e-16)
        r1 = utils.reflect(r1)
        r2 = F(r1)
        r2 = utils.reflect(r2)
        RL = RL / (np.abs(r2)+1e-16)

        r1 = y / (np.abs(F(RL))+1e-16)
        r1 = utils.reflect(r1)
        r2 = F(r1)
        r2 = utils.reflect(r2)
        RL = RL * r2

    return RL

